### PR TITLE
Add IQueryTypeConfiguration<TEntity>

### LINKED ISSUE #11257: Feature parity: Add IQueryTypeConfiguration<TEntity>
I really like the addition of the new `IEntityTypeConfiguration<TEntity>` type that allows us [self-contained type configuration for code first](https://docs.microsoft.com/en-us/ef/core/what-is-new/ef-core-2.0#self-contained-type-configuration-for-code-first) and if I'm not missing anything, this types allows only entity (i.e. not query types) to be configured. 
I think ef can provide the same experience for query types as well so I propose adding a new `IQueryTypeConfiguration<TEntity>` with this shape:

```csharp
public interface IQueryTypeConfiguration<TEntity>
   where TEntity : class
{
     void Configure(QueryTypeBuilder<TEntity> builder);
}
```
 as well as a new `ApplyConfiguration` overload on the `ModelBuilder` class like the following:

```csharp
public virtual ModelBuilder ApplyConfiguration<TEntity>([NotNull] IQueryTypeConfiguration<TEntity> configuration)
   where TEntity : class
{
   // implementation omitted
}
```
allowing us to write configuration for entity types and query types in a consistent way.

### PR DESCRIPTION
**Pull request by ilmax**
**Please check if the PR fulfills these requirements**

- [x] The code builds and tests pass (verified by our automated build checks)
- [ ] Commit messages follow this format
```
    Summary of the changes
    - Detail 1
    - Detail 2

```
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Code meets the expectations our engineering guidelines. https://github.com/aspnet/Home/wiki/Engineering-guidelines.

Fixes #11257

This is my first PR for EF, so please be patient with me 😃

### CODE REVIEW

File: src/EFCore/IQueryTypeConfiguration`.cs [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+    ///     Allows configuration for a query type to be factored into a separate class,
+    ///     rather than in-line in <see cref="DbContext.OnModelCreating(ModelBuilder)" />.
+    ///     Implement this interface, applying configuration for the entity in the
+    ///     <see cref="Configure(QueryTypeBuilder{TEntity})" /> method,
+    ///     and then apply the configuration to the model using
+    ///     <see cref="ModelBuilder.ApplyConfiguration{TEntity}(IQueryTypeConfiguration{TEntity})" />
+    ///     in <see cref="DbContext.OnModelCreating(ModelBuilder)" />.
+    /// </summary>
+    /// <typeparam name="TEntity"> The query type to be configured. </typeparam>
+    public interface IQueryTypeConfiguration<TEntity>
```
Comment (anpete) [inline src/EFCore/IQueryTypeConfiguration`.cs]: Change uses of 'entity' to query type (generic and doc comments)

File: test/EFCore.Tests/ModelBuilding/IQueryTypeConfigurationTest.cs [hunk 2]
```diff
@@ -0,0 +1,89 @@
+﻿using System.Linq;
+using Microsoft.EntityFrameworkCore.Metadata.Builders;
+using Microsoft.EntityFrameworkCore.TestUtilities;
+using Xunit;
+
+namespace Microsoft.EntityFrameworkCore.ModelBuilding
+{
+    public abstract partial class ModelBuilderTest
+    {
+
```
Comment (anpete) [inline test/EFCore.Tests/ModelBuilding/IQueryTypeConfigurationTest.cs]: nit: whitespace

File: src/EFCore/IQueryTypeConfiguration`.cs (hunk 1 above)
Comment (ilmax) [reply src/EFCore/IQueryTypeConfiguration`.cs]: also the generic parameter?

Comment (anpete) [reply src/EFCore/IQueryTypeConfiguration`.cs]: Yes: TQuery

File: src/EFCore/IQueryTypeConfiguration`.cs [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+
+namespace Microsoft.EntityFrameworkCore
+{
+    /// <summary>
+    ///     Allows configuration for a query type to be factored into a separate class,
+    ///     rather than in-line in <see cref="DbContext.OnModelCreating(ModelBuilder)" />.
+    ///     Implement this interface, applying configuration for the query in the
+    ///     <see cref="Configure(QueryTypeBuilder{TQuery})" /> method,
+    ///     and then apply the configuration to the model using
+    ///     <see cref="ModelBuilder.ApplyConfiguration{TEntity}(IQueryTypeConfiguration{TEntity})" />
```
Comment (smitpatel) [inline src/EFCore/IQueryTypeConfiguration`.cs]: `TQuery` ?

Comment (ilmax) [reply src/EFCore/IQueryTypeConfiguration`.cs]: Damn :( will fix!

**Comment by ilmax** [comment]
Should I rebase it?

**Comment by anpete** [comment]
@ilmax Yes, thanks!

**Comment by ilmax** [comment]
@anpete Done, let me know if something is missing!

**Comment by ilmax** [comment]
Any idea about build failures? To me it seems unrelated to the change

**Comment by smitpatel** [comment]
@ilmax - I have triggered appveyor/travis builds again. Rest of the CIs are passing so its good sign.
Travis is hitting intermittent issue with SQLite - database is locked error.
AppVeyor failed because we had local build break.

**Comment by smitpatel** [comment]
Also can you rebase on release/2.1 branch?

**Comment by ilmax** [comment]
Yes, will do tomorrow CET time

**Comment by anpete** [comment]
Merged thanks!

**Pull request merged**